exec/prepare-package/src/index.ts

```diff
@@ -62,3 +64,16 @@ function packageShouldBeBuilt (manifest: PackageManifest, pkgDir: string): boole
   const mainFile = manifest.main ?? 'index.js'
   return !fs.existsSync(path.join(pkgDir, mainFile))
 }
+
+function safeJoinPath (root: string, sub: string) {
+  const joined = path.join(root, sub)
+  // prevent the dir traversal attack
+  const relative = path.relative(root, joined)
+  if (relative.startsWith('..')) {
```

Comment:
Doesn't this still allow passing `subdir/../../../` and an `relative.includes('..')` would be safer?

Reply:
`relative` is a calculated relative path between two ABSOLUTE paths: `root` and `joined`. If you pass `subdir/../../../`, the `relative` will definitely start with `..`.

Reply:
Oh. I missed the call to `path.relative` before.

Reply:
We should always use `path.relative` to let Node.js calculate whether the user input path is outside the project root. Disallowing `..` or `/` might not be a safe way since the resolving rules may change in future, and our block list may not be effective then :)